Move image pull abilities to CDH (#11197)
kata-containers/kata-containers: docs/design/kata-guest-image-management-design.md

#### Architecture

The following diagram provides an overview of the architecture for pulling image in the guest with key components. 
```mermaid
flowchart LR
    Kubelet[kubelet]--> |1 Pull image request & metadata|Containerd

>     Containerd-->Snapshotter[Nydus Snapshotter]

Reviewer: Do we want to keep the `[Container Images Registry]-->|2. Pull image metadata| Containerd` to reflect that there is still a host side pull that is needed?

Author: The render of `mermaid` would meet some trouble when I used `.`. Let me try to resolve this.

>     Kubelet[kubelet]--> |1 Pull image request & metadata|Containerd
>     Containerd-->Snapshotter[Nydus Snapshotter]
>     Snapshotter-->|2 pack image info| Containerd
>     Containerd-->Runtime[Kata Runtime]
>     Runtime-->Hypervisor
>     Hypervisor-->TEE
>     Runtime-->|3 pass image info to VM| Agent
>     E[Container Images Registry]-->|5 pull image with image info|CDH1
>     subgraph TEE [Virtual Machine]
>       Images[Container Images]-->|6 prepare container rootfs|H[Container]
> 
>       subgraph CDH [Confidential Data Hub]
>         CDH1[Image Mgmt]
>       end
> 
>       CDH-->Images
>       Agent[Kata Agent]-->|4 call image pull RPC|CDH

Reviewer: nit: Add a period after each number and capitalise for readability
```suggestion
    Kubelet[kubelet]--> |1. Pull image request & metadata|Containerd
    Containerd-->Snapshotter[Nydus Snapshotter]
    Snapshotter-->|2. Pack image info| Containerd
    Containerd-->Runtime[Kata Runtime]
    Runtime-->Hypervisor
    Hypervisor-->TEE
    Runtime-->|3. Pass image info to VM| Agent
    E[Container Images Registry]-->|5. Pull image with image info|CDH1
    subgraph TEE [Virtual Machine]
      Images[Container Images]-->|6. Prepare container rootfs|H[Container]

      subgraph CDH [Confidential Data Hub]
        CDH1[Image Mgmt]
      end

      CDH-->Images
      Agent[Kata Agent]-->|4. Call image pull RPC|CDH
```

Author: The render of `mermaid` would meet some trouble when I used `.`. Let me try to resolve this.

Reviewer: Ok, no problem if there is a mermaid issue. I think the capitalisation will help with the distinction

Author: Fixed, thanks

#### Architecture

The following diagram provides an overview of the architecture for pulling image in the guest with key components. 
```mermaid
flowchart LR
    Kubelet[kubelet]--> |1\. Pull image request & metadata|Containerd

>     E-->|2\. Pull image metadata| Containerd

Reviewer 2: Should this arrow be going _from_ containerd _to_ the registry?

    Containerd-->Snapshotter[Nydus Snapshotter]
    Snapshotter-->|3\. Pack image info| Containerd
    Containerd-->Runtime[Kata Runtime]
    Runtime-->Hypervisor
    Hypervisor-->TEE
    Runtime-->|4\. Pass image info to VM| Agent

>     E[Container Images Registry]-->|6\. Pull image with image info|CDH1

Reviewer 2: I'd also expect this arrow to go the opposite direction.

    subgraph TEE [Virtual Machine]
      Images[Container Images]-->|7\. Prepare container rootfs|H[Container]

      subgraph CDH [Confidential Data Hub]
        CDH1[Image Mgmt]
      end

      CDH-->Images
      Agent[Kata Agent]-->|5\. Call image pull RPC|CDH
    end
```

#### Sequence diagrams

The following sequence diagram depicted below offers a detailed overview of the messages/calls exchanged to pull an unencrypted unsigned image from an unauthenticated container registry. This involves the kata-runtime, kata-agent, and the guest-components’ image-rs to use the guest pull mechanism.

```mermaid
sequenceDiagram
    par Hosts Side
        Containerd/Kubelet->>runtime.kata_agent: createContainer(ctx,sandbox,c)
        runtime.kata_agent-->runtime.fs_share_linux: ShareRootFilesystem(ctx,c)

>         runtime.fs_share_linux->>runtime.kata_agent: ShandleVirtualVolumeStorageObject(c,...,KataVolumeType)

Reviewer 2: ```suggestion
        runtime.fs_share_linux->>runtime.kata_agent: handleVirtualVolumeStorageObject(c,...,KataVolumeType)
```

>         runtime.fs_share_linux->>runtime.kata_agent: ShandleVirtualVolumeStorageObject(c,...,KataVolumeType)

Reviewer 2: This sequence looks strange: agent calls fs_share, and then fs_share calls back agent? If this is indeed what happens, I think it would benefit from an explanation.

Author: Yes. This sequence follows what it is today https://github.com/kata-containers/kata-containers/blob/main/docs/design/kata-guest-image-management-design.md#sequence-diagrams

Reviewer 2: Thanks, it's easier to understand now with the adjusted arrows.

        runtime.kata_agent-->runtime.kata_agent: handleImageGuestPullBlockVolume(c,virtVolume,vol)
        runtime.kata_agent-->runtime.fs_share_linux: ret:storage

>         runtime.fs_share_linux->>runtime.kata_agent: ret:sharedFile

Reviewer 2: Also not quite clear what's happening here: the `-->` arrows don't have a head - what does that mean for the control flow?

Author: Fixed with all head

    and Guest Side
        runtime.kata_agent->>agent.rpc: CreateContainerRequest(cid,...,storages,...,oci,...)
        agent.rpc->>agent.storage: add_storage(storages...)
        agent.storage->>agent.storage: StorageHandler.handler(driver)
        agent.storage->>agent.storage.StorageHandler.ImagePullHandler: create_device(storage)
        agent.storage.StorageHandler.ImagePullHandler->>agent.confidential_data_hub: pull_image(img,cid,img_metadata)
        agent.confidential_data_hub->>Confidential Data Hub: pull_image(img,bundle_path)

>         Confidential Data Hub->>agent.confidential_data_hub: ret: manifest_digest

Reviewer 2: Nit after updating the proto:

```suggestion
        Confidential Data Hub->>agent.confidential_data_hub: ret
```

```
Next, the kata-agent's RPC module will handle the create container request which, among other things, involves adding storages to the sandbox. The storage module contains implementations of `StorageHandler` interface for various storage types, being the `ImagePullHandler` in charge of handling the storage object for the container image (the storage manager instantiates the handler based on the value of the "driver").

> `ImagePullHandler` delegates the image pulling operation to the `confidential_data_hub.pull_image()` that is going to create the image's bundle directory on the guest filesystem and, in turn, class Confidential Data Hub to in fact fetch and uncompress the image's bundle. 

Reviewer 2: ```suggestion
`ImagePullHandler` delegates the image pulling operation to the `confidential_data_hub.pull_image()` that is going to create the image's bundle directory on the guest filesystem and, in turn, the `ImagePullService` of Confidential Data Hub to fetch, uncompress and mount the image's rootfs. 
```

References:
[1] [[RFC] Image management proposal for hosting sharing and peer pods](https://github.com/confidential-containers/confidential-containers/issues/137)
[2] https://github.com/containerd/containerd/blob/main/docs/content-flow.md

> [3] [Move guest pull ability to a configurable component](https://github.com/kata-containers/kata-containers/issues/9266)

Reviewer 2: While you're at it, you could improve the references section by making it a numbered list:

```suggestion
1. [[RFC] Image management proposal for hosting sharing and peer pods](https://github.com/confidential-containers/confidential-containers/issues/137)
2. https://github.com/containerd/containerd/blob/main/docs/content-flow.md
3. [Move guest pull ability to a configurable component](https://github.com/kata-containers/kata-containers/issues/9266)
```
